mantis-connectors/mantis-connector-iceberg/src/main/java/io/mantisrx/connector/iceberg/sink/committer/IcebergCommitter.java

```diff
@@ -16,36 +16,82 @@
+import static io.mantisrx.connector.iceberg.sink.writer.DefaultIcebergWriter.minNullSafe;
+
+import io.mantisrx.connector.iceberg.sink.committer.config.CommitterConfig;
+import io.mantisrx.connector.iceberg.sink.writer.MantisDataFile;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import javax.annotation.Nullable;
+import lombok.extern.slf4j.Slf4j;
 import org.apache.iceberg.AppendFiles;
 import org.apache.iceberg.DataFile;
 import org.apache.iceberg.Table;
+import org.apache.iceberg.Transaction;
+import org.apache.iceberg.UpdateProperties;
 
 
 /**
  * Commits {@link DataFile}s for Iceberg tables.
  *
  * This class uses Iceberg's Table API and only supports Table#append operations.
  */
+@Slf4j
 public class IcebergCommitter {
 
     private final Table table;
+    private final CommitterConfig config;
 
-    public IcebergCommitter(Table table) {
+    public IcebergCommitter(Table table, CommitterConfig committerConfig) {
         this.table = table;
+        this.config = committerConfig;
     }
 
     /**
      * Uses Iceberg's Table API to append DataFiles and commit metadata to Iceberg.
      *
      * @return the current snapshot of the table.
      */
-    public Map<String, Object> commit(List<DataFile> dataFiles) {
-        AppendFiles tableAppender = table.newAppend();
-        dataFiles.forEach(tableAppender::appendFile);
+    public Map<String, Object> commit(List<MantisDataFile> dataFiles) {
+        Transaction transaction = table.newTransaction();
+
+        AppendFiles tableAppender = transaction.newAppend();
+        dataFiles.stream().map(MantisDataFile::getDataFile).forEach(tableAppender::appendFile);
         tableAppender.commit();
+        log.info(
+            "Iceberg committer {}.{} appended {} data files to transaction",
+            config.getDatabase(),
+            config.getTable(),
+            dataFiles.size());
+
+        Long currentWatermark = getCurrentWatermark(transaction.table());
+        Long lowWatermark = null;
+        for (MantisDataFile flinkDataFile : dataFiles) {
+            lowWatermark = minNullSafe(lowWatermark, flinkDataFile.getLowWatermark());
+        }
+        final Long finalWatermark = maxNullSafe(currentWatermark, lowWatermark);
+
+        if (finalWatermark != null) {
+            UpdateProperties updateProperties = transaction.updateProperties();
+            updateProperties.set(config.getWatermarkPropertyKey(), Long.toString(finalWatermark));
+            updateProperties.commit();
+            log.info("Iceberg committer for table={} set VTTS watermark to {}", config.getTable(), finalWatermark);
+        }
+
+
+        transaction.commitTransaction();
         return table.currentSnapshot() == null ? new HashMap<>() : new HashMap<>(table.currentSnapshot().summary());
     }
+
+    @Nullable
+    private Long getCurrentWatermark(Table table) {
+        try {
+            return Long.parseLong(table.properties().get(config.getWatermarkPropertyKey()));
```

Comment:
I'm thinking if we need to refresh table? Otherwise the `table.properties()` may always return the old value. I'm not sure about this though. A safer approach is to create a `lowWatermark` field in this committer class, initialize it once in the constructor, and keep updating it for every commit. Then we don't need this `getCurrentWatermark` method any more as it's saved and maintained in the Iceberg committer class.

Reply:
Since we are sending the table from the current transaction, it should have the latest set of properties.